pkg/bitmap/bitmap.go

```diff
@@ -0,0 +1,232 @@
+	}
+	beginBlock := begin / 64
+	endBlock := end / 64
+	if beginBlock == endBlock {
+		b.bitBlock[endBlock] ^= ((^uint64(0) << uint(begin%64)) & ((uint64(1) << (uint(end) % 64 + 1))-1))
+	} else {
+		b.bitBlock[beginBlock] ^= ^(^uint64(0) << uint(begin%64))
+		for i := beginBlock; i < endBlock; i++ {
+			b.bitBlock[i] = ^b.bitBlock[i]
+		}
+		b.bitBlock[endBlock] ^= ((uint64(1) << (uint(end) % 64 + 1))-1)
+	}
+}
+
+// removeBitmapRange remove the bits within range (begin and end). begin and end are inclusive.
+func (b *Bitmap) removeBitmapRange(begin, end int) {
+	if begin > end {
+		return
+	}
+	if begin == end {
+		blockNum, mask := begin/64, uint64(1)<<(begin%64)
+		b.bitBlock[blockNum] = b.bitBlock[blockNum] &^ mask
+		return
+	}
+	beginBlock := begin / 64
+	endBlock := end / 64
+	if beginBlock == endBlock {
+		b.bitBlock[beginBlock] &= (((uint64(1) << uint(begin%64)) - 1) | ^((uint64(1) << (uint(end) % 64 + 1)) - 1))
+	} else {
+		b.bitBlock[beginBlock] &= ((uint64(1) << uint(begin%64)) - 1)
+		for i := beginBlock + 1; i < endBlock; i++ {
+			b.bitBlock[i] &= ^b.bitBlock[i]
+		}
+		b.bitBlock[endBlock] &= ^((uint64(1) << (uint(end) % 64 + 1)) - 1)
+	}
+}
+
+// RemoveRange remove bits within range (begin and end) for the Bitmap. begin and end are inclusive.
+func (b *Bitmap) RemoveRange(begin, end int) {
+	blockRange := end/64 - begin/64
+	// when the number of removed blocks is larger than half of the length of b.bitBlock
+	// we get the length by countOnesForAllBlocks for better performance
+	if blockRange > len(b.bitBlock)/2 {
+		b.removeBitmapRange(begin, end)
+		b.numOnes = int(b.countOnesForAllBlocks())
+	} else {
+		oldRangeOnes := b.countOnesForBlocks(begin, end)
+		b.removeBitmapRange(begin, end)
+		newRangeOnes := b.countOnesForBlocks(begin, end)
+		b.numOnes += int(newRangeOnes - oldRangeOnes)
+	}
+}
+
+// Flip flip bits within range (begin and end) for the Bitmap. begin and end are inclusive.
+func (b *Bitmap) Flip(begin, end int) {
+	blockRange := end/64 - begin/64
+	// when the number of flipped blocks is larger than half of the length of b.bitBlock
+	// we get the length by countOnesForAllBlocks for better performance
+	if blockRange > len(b.bitBlock)/2 {
+		b.flipBitmapRange(begin, end)
+		b.numOnes = int(b.countOnesForAllBlocks())
+	} else {
+		oldRangeOnes := b.countOnesForBlocks(begin, end)
+		b.flipBitmapRange(begin, end)
+		newRangeOnes := b.countOnesForBlocks(begin, end)
+		b.numOnes += int(newRangeOnes - oldRangeOnes)
+	}
+}
+
+// ToSlice transform the Bitmap into slice.
+func (b *Bitmap) ToSlice() []int32 {
+	bitmapSlice := make([]int32, b.numOnes, b.numOnes)
+	// base is the start number of a bitBlock
+	base := 0
+	// index is the index of bitmapSlice
+	index := 0
+	for i := 0; i < len(b.bitBlock); i++ {
+		bitBlock := b.bitBlock[i]
+		for bitBlock != 0 {
+			j := bitBlock & -bitBlock
+			bitmapSlice[index] = int32((base + int(bits.OnesCount64(j - 1))))
+			index = index + 1
+			bitBlock ^= j
+		}
+		base += 64
```

Comment:
there is some bit gymnastics going on here, might be worth adding a comment like:
```
// Iterate through all the numbers held by this bit block.
for bitBlock != 0 {
...
// Extract the LSB.
j := bitBlock & -bitBlock
// Interpret the bit as the uin32 number it represents and add it to result.
ret = append(ret, int32((base + int(bits.OnesCount64(j - 1)))))